File: IAIso-v5.0/core/iaiso-ruby/lib/iaiso/consent/verifier.rb
# frozen_string_literal: true

require_relative "algorithm"
require_relative "errors"
require_relative "jwt"
require_relative "scope"

module IAIso
  module Consent
    # Verifies signed consent tokens.
    class Verifier
      attr_reader :algorithm, :issuer, :leeway_seconds

      def initialize(algorithm: Algorithm::HS256, issuer: "iaiso",
                     hs_key: nil, rs_public_key: nil,
                     revocation_list: nil, leeway_seconds: 5, clock: nil)
        @algorithm = algorithm
        @issuer = issuer
        @hs_key = hs_key
        @rs_public_key = rs_public_key
        @revocation_list = revocation_list
        @leeway_seconds = leeway_seconds.to_i
        @clock = clock || ->{ Time.now.to_i }
      end

      # Verify `token`. If `requested_execution_id` is non-nil and the
      # token is bound to a different execution, raises InvalidTokenError.
      def verify(token, requested_execution_id: nil)
        parsed = JWT.parse(token)

        alg = parsed[:header]["alg"]
        unless alg == @algorithm
          raise InvalidTokenError, "unexpected algorithm: #{alg.inspect}"
        end

        ok = JWT.verify_signature(
          parsed,
          algorithm: @algorithm,
          hs_key: @hs_key,
          rs_public_key: @rs_public_key,
        )
        raise InvalidTokenError, "signature verification failed" unless ok

        claims = parsed[:claims]
        %w[exp iat iss sub jti].each do |k|
          unless claims.key?(k)
            raise InvalidTokenError, "missing required claim: #{k}"
          end
        end

        unless claims["iss"] == @issuer
          raise InvalidTokenError, "iss mismatch: got #{claims["iss"].inspect}, want #{@issuer.inspect}"
        end

        exp = claims["exp"].to_i
        now = @clock.call.to_i
        if exp + @leeway_seconds < now
          raise ExpiredTokenError
        end

        jti = claims["jti"].to_s
        if @revocation_list && @revocation_list.revoked?(jti)
          raise RevokedTokenError, jti
        end

        token_exec = claims["execution_id"]
        if requested_execution_id && token_exec && token_exec != requested_execution_id
          raise InvalidTokenError,
                "token bound to #{token_exec.inspect}, requested #{requested_execution_id.inspect}"
        end

        scopes = (claims["scopes"] || []).map(&:to_s)
        metadata = claims["metadata"].is_a?(Hash) ? claims["metadata"] : {}

        Scope.new(
          token: token,
          subject: claims["sub"].to_s,
          scopes: scopes,
          execution_id: token_exec,
          jti: jti,
          issued_at: claims["iat"].to_i,
          expires_at: exp,
          metadata: metadata,
        )
      end
    end
  end
end

File: IAIso-v5.0/core/iaiso-swift/Sources/IAIsoConsent/Verifier.swift
import Foundation

/// Verifies signed consent tokens.
public final class Verifier: @unchecked Sendable {
    public let algorithm: Algorithm
    public let issuer: String
    public let leewaySeconds: Int64
    private let hsKey: Data?
    private let rsPublicKey: Data?
    private let revocationList: RevocationList?
    private let clock: () -> Int64

    public init(
        algorithm: Algorithm = .hs256,
        issuer: String = "iaiso",
        hsKey: Data? = nil,
        rsPublicKey: Data? = nil,
        revocationList: RevocationList? = nil,
        leewaySeconds: Int64 = 5,
        clock: @escaping () -> Int64 = { Int64(Date().timeIntervalSince1970) }
    ) {
        self.algorithm = algorithm
        self.issuer = issuer
        self.hsKey = hsKey
        self.rsPublicKey = rsPublicKey
        self.revocationList = revocationList
        self.leewaySeconds = leewaySeconds
        self.clock = clock
    }

    /// Verify `token`. If `requestedExecutionId` is non-nil and the token
    /// is bound to a different execution, throws `ConsentError.invalidToken`.
    public func verify(
        _ token: String,
        requestedExecutionId: String? = nil
    ) throws -> Scope {
        let parsed = try JWT.parse(token)

        // Algorithm check from header.
        guard let alg = parsed.header["alg"] as? String else {
            throw ConsentError.invalidToken(reason: "missing algorithm")
        }
        guard alg == algorithm.rawValue else {
            throw ConsentError.invalidToken(reason: "unexpected algorithm: \(alg)")
        }

        // Signature.
        let valid = try JWT.verifySignature(
            parsed,
            algorithm: algorithm,
            hsKey: hsKey,
            rsPublicKey: rsPublicKey)
        guard valid else {
            throw ConsentError.invalidToken(reason: "signature verification failed")
        }

        let claims = parsed.claims
        for req in ["exp", "iat", "iss", "sub", "jti"] {
            if claims[req] == nil {
                throw ConsentError.invalidToken(reason: "missing required claim: \(req)")
            }
        }

        guard let iss = claims["iss"] as? String, iss == issuer else {
            let got = (claims["iss"] as? String) ?? "<missing>"
            throw ConsentError.invalidToken(
                reason: "iss mismatch: got \(got), want \(issuer)")
        }

        let exp = numericClaim(claims["exp"]) ?? 0
        let now = clock()
        if exp + leewaySeconds < now {
            throw ConsentError.expiredToken
        }

        let jti = (claims["jti"] as? String) ?? ""
        if let rl = revocationList, rl.isRevoked(jti) {
            throw ConsentError.revokedToken(jti: jti)
        }

        let tokenExec = claims["execution_id"] as? String
        if let req = requestedExecutionId, let bound = tokenExec, bound != req {
            throw ConsentError.invalidToken(
                reason: "token bound to \(bound), requested \(req)")
        }

        let subject = (claims["sub"] as? String) ?? ""
        let iat = numericClaim(claims["iat"]) ?? 0
        let scopes = (claims["scopes"] as? [String]) ?? []
        var metadata: [String: String] = [:]
        if let m = claims["metadata"] as? [String: Any] {
            for (k, v) in m {
                if let s = v as? String { metadata[k] = s }
            }
        }
        return Scope(
            token: token,
            subject: subject,
            scopes: scopes,
            executionId: tokenExec,
            jti: jti,
            issuedAt: iat,
            expiresAt: exp,
            metadata: metadata)
    }

    private func numericClaim(_ v: Any?) -> Int64? {
        if let n = v as? NSNumber { return n.int64Value }
        if let i = v as? Int64 { return i }
        if let i = v as? Int { return Int64(i) }
        if let d = v as? Double { return Int64(d) }
        return nil
    }
}
